### ISSUE TITLE
Cancelling a timer promise

### ISSUE THREAD

**Issue by barakp** [sandstorm-io/capnproto#263]
Hi @kentonv ,
What would be the best approach to cancel a detached timer promise? currently I have this ugly hack that stores a shared pointer in the `then` lambda context that's being set when cancelling the timer and checked upon the call to the lambda. I'm sure there's a better way to do that but was unable to find it in the API. Assuming detached promises can't be cancelled (as they're moved into capnp), is there a better way to cancel a timer promise silently (i.e. without rejecting the promise but also without fulfilling it)?

Your kind advice would be much appreciated.

**Comment by kentonv** [comment]
Hi Barak,

Generally I recommend against detaching promises. In well-written code there's usually a good place to store the promise, e.g. some object that represents whatever job the background task is a part of. If you haven't detached the promise then you can cancel it by destroying it (assign it to null, or let it go out of scope).

Another strategy that can work is to take advantage of `exclusiveJoin()`. Create a PromiseFulfillerPair, then join the promise part together with your timer promise. You can then fulfill (or reject) the fulfiller to cancel the wait, because `exclusiveJoin()` resolves as soon as either input resolves (and cancels the other input). Note that with this strategy you have to keep in mind that when you fulfill the canceller promise, the timer event may already be queued and so the callback may still be called as if no cancellation occurred. In contrast, when you cancel a promise by destroying it, you have the guarantee that no further callbacks related to that promise will execute.

**Comment by barakp** [comment]
Hi Kenton,

Thanks for your prompt reply. I was trying to avoid another single call to the callback after the timer is cancelled because the object that the callback refers to may already be destroyed. I only took the detach approach because I found it as the most convenient way in capnp to schedule a callback on the underlying event loop. I was under the impression that the only two ways to get the callback called are either eventually calling wait() on the promise using the underlying WaitScope or calling detach. Is there any other way to register the promise on the event loop other than calling wait and/or detach?

**Comment by kentonv** [comment]
Ah, the problem you're running into is that `.then()` continuations which aren't themselves asynchronous (i.e. they return a value, not a promise) are lazily executed. There are several ways to force a continuation to execute eagerly:
* `.wait()` on it.
* `.detach()` it.
* Add it to a `kj::TaskSet` (this is a lot like `detach()`, except that you can cancel all tasks in the set by destroying the set).
* Call `.eagerlyEvaluate()` on it.

**Comment by barakp** [comment]
Yep - that was just it. Thank you!
